pkg/services/ngalert/state/historian/loki_http.go

```diff
@@ -25,13 +25,22 @@ func NewRequester() client.Requester {
 	}
 }
 
+// encoder serializes log streams to some byte format.
+type encoder interface {
+	// encode serializes a set of log streams to bytes.
+	encode(s []stream) ([]byte, error)
+	// headers returns a set of HTTP-style headers that describes the encoding scheme used.
+	headers() map[string]string
+}
+
```

Comment:
nit: you're using the concrete implementation in tests anyways I understand the "programming practice" but YAGNI it's a thing.

Reply:
We already do need it though - the main path uses snappyProto (since that's the main encoding) and tests use json (since it's human readable) so even within this PR we're already swapping out the implementation.

I originally tried it without the interface - enum for encoding type and if statements to call the right encoding method. The result felt pretty bug prone to me, as we needed branches in a number of places to get it right.

Reply:
Makes sense, I missed that detail!